### ISSUE TITLE
Support for negative segment sizes

### ISSUE THREAD

**Issue by senior-zero** [NVIDIA/cub#593]
In our segmented reduce we state:
>  If `d_end_offsets[i] - 1 <= d_begin_offsets[i]`, the *i*<sup>th</sup> segment is considered empty.
In our segmented sort and segmented radix sort we state the same thing. 

While in the segmented reduce implementation, we check that:
```cpp
if (segment_begin == segment_end)
```
to recognize empty segments. I don't think we test for negative segment sizes. 

I suggest we decide if we want to support negative offsets and then add necessary tests / fixes, or we change the docs.

**Comment by miscco** [comment]
I am abhorred by the thought of someone passing a negative size intentionally, but it might happen accidentally due to some integer shenanigance. 

So generally I would always prefer to defensively check `<=` rather than `==` for a lower boundary.

I just recently had a discussion about this in some cursed string function where it was an actual bug

**Comment by senior-zero** [comment]
@miscco, I agree and also think that it's safer to fix the code than to fix the docs and let people find out that their code is silently broken.
